src/cockpit/safely.py

```diff
@@ -0,0 +1,74 @@
+# This file is part of Cockpit.
+#
+# Copyright (C) 2023 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import contextlib
+
+from typing import Callable, Dict, Generator, List, Mapping, Type, TypeVar, Union
+
+T = TypeVar('T')
+D = TypeVar('D')
+
+
+@contextlib.contextmanager
+def catch(description: str, on_error: Callable) -> Generator[None, None, None]:
+    try:
+        yield
+    except (TypeError, KeyError) as err:
+        on_error(f'{description}: {err}')
+        raise
+
+
+def get(mapping: Mapping[str, object], key: str, expected_type: Type[T], *default: D) -> Union[T, D]:
+    try:
+        value = mapping[key]
+    except KeyError:
+        if default:
+            return default[0]
```

Comment:
`*default` means that you can add arbitrarily many extra arguments, which will just be ignored. This is reflected here in this rather ugly `[0]`. IMHO this should rather be declared as a kwarg `default: Optional[T]=None`?

Same for get_{list,dict} below.

Reply:
The issue is that I want to be able to detect the difference between no default value being given (in which case it's a hard error if the key lookup fails) and a default value of `None` being given.

I'd do two separate functions for this, but then I'd need to build two completely separate hierarchies...

The other option I considered was using a different sentinel value — maybe that would indeed be better.  That presents a problem with getting the types working nicely, though.